### PR TITLE
doc: genrest: Use Kconfiglib 2 and improve output

### LINKED ISSUE #5622: Update kconfiglib used for generating Kconfig documentation
We do have an old version that needs to be updated and use the new one being imported into scripts/

### LINKED ISSUE #6821: Simplify rendering of Kconfig variable doc
Currently Kconfig options are rendered in an excessively verbose way:

http://docs.zephyrproject.org/reference/kconfig/CONFIG_NET_DEBUG_L2_IEEE802154.html#cmdoption-arg-config-net-debug-l2-ieee802154

Propose to reduce that to smaller set of variables such as:

* Title
* Help text
* Dependencies
* Value
* File defined

### PR DESCRIPTION
**Pull request by ulfalizer**
The new version of the reference page for a symbol lists the following
essential and generally useful information at the top of the page:

 - Prompt

 - Type

 - Help text

 - Direct dependencies (from 'depends on', and inherited from if's and
   menus)

 - Defaults

 - Other symbols selecting or implying the symbol

The current symbol value and visibility are no longer displayed. They
don't make much sense in a reference document, as they depend on user
values from configuration files.

The bottom of the reference page also includes the symbol represented in
Kconfig format (including propagated dependencies).

All references to defined symbols appearing in expressions are turned
into links, which can be clicked on to jump to the reference page for
that symbol.

The following (slightly outdated) screenshot gives an idea of the
format:
https://www.dropbox.com/s/a34tlk2ncyus8po/promptandtype.png?dl=0

This commit also removes Kconfiglib 1, since it is no longer used.

Fixes #5622
Fixes #6821

Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

### CODE REVIEW

**Comment by mbolivar** [comment]
> All references to defined symbols appearing in expressions are turned
> into links, which can be clicked on to jump to the reference page for
> that symbol.

Does this include symbols selected by a given config?

**Comment by ulfalizer** [comment]
@mbolivar 
I was just going to create a separate issue for that. It is broken at the moment, but will be fixed once https://github.com/ulfalizer/Kconfiglib/commit/a5ea0dcbc8b9343f6901654ad92de6ece156f954 gets in.

I also want to have `A && B || C && D` rendered as `(A && B) || (C && D)` for clarity (bit silly to require people to memorize the precedence table). That'll be rolled into the next Kconfiglib update as well.

**Comment by ulfalizer** [comment]
@mbolivar https://github.com/zephyrproject-rtos/zephyr/issues/6844

File: doc/scripts/genrest/genrest.py [hunk 1]
```diff
@@ -96,15 +63,148 @@ def print_items(items, outdir, indent):
...
+                              kconfiglib.TYPE_TO_STR[sym.type]))
+
+        # Symbols with multiple definitions can have multiple help texts
+        for node in sym.nodes:
+            if node.help is not None:
+                # Wrap the help text in a code block to preserve "ASCII art"
+                # and the like
+                sym_rst.write("Help\n"
+                              "====\n\n"
+                              ".. code-block:: none\n\n"
```
Comment (dbkinder) [inline doc/scripts/genrest/genrest.py]: I think there are folks using markup tags in the help text for lists and such that won't render as expected if treated as a code block.

Comment (ulfalizer) [reply doc/scripts/genrest/genrest.py]: @dbkinder 
Not sure about Zephyr, but I've seen quite a lot of help texts with figures drawn in text in the Linux kernel.

I agree that the output is uglier compared to a plain paragraph.

**Comment by ulfalizer** [comment]
Another slightly wonky thing that's worth mentioning:

If a symbol is defined in multiple locations, all the properties that are tied to the symbol itself (the `default`s, `select`s, `imply`s, and `range`s), as opposed to being tied to the menu node (e.g. `prompt` and `help`), will be displayed on the first symbol definition in the Kconfig display. It's tricky to work around, as that's how Kconfig sees things internally.

I wonder if that's confusing enough to be worth mentioning on the reference pages.

File: doc/scripts/genrest/genrest.py (hunk 1 above)
Comment (ulfalizer) [reply doc/scripts/genrest/genrest.py]: Oh, you mean they use RST formatting in help texts? Yeah, in that case, it's definitely wrong to wrap it in a `code-block`.

File: doc/scripts/genrest/genrest.py [hunk 2]
```diff
@@ -84,9 +48,12 @@ def print_items(items, outdir, indent):
 
 The configuration options' information below is extracted directly from
 :program:`Kconfig` using the :file:`~/doc/scripts/genrest/genrest.py` script.
-Click on the option name in the table below for detailed information about
-each option.
+Click on the symbol name in the table below for to see its definition.
```
Comment (ulfalizer) [inline doc/scripts/genrest/genrest.py]: Heh... not sure why I changed this. Leftover from some earlier rewriting I think. I could change it back.

**Comment by ulfalizer** [comment]
> Another slightly wonky thing that's worth mentioning:
> 
> If a symbol is defined in multiple locations, all the properties that are tied to the symbol itself (the defaults, selects, implys, and ranges), as opposed to being tied to the menu node (e.g. prompt and help), will be displayed on the first symbol definition in the Kconfig display. It's tricky to work around, as that's how Kconfig sees things internally.
> 
> I wonder if that's confusing enough to be worth mentioning on the reference pages.

Just remembered that I mentioned it in a new paragraph in `index.rst`.

File: doc/scripts/genrest/genrest.py (hunk 1 above)
Comment (ulfalizer) [reply doc/scripts/genrest/genrest.py]: Talked a bit to @mbolivar on IRC. It wouldn't take anything fancy for the formatting to get messed up either. Take a look at `SYS_LOG_SPI_LEVEL` in `drivers/spi/Kconfig` for example.

Looks like you can preserve newlines in RST by putting `|` at the beginning of each line. There are probably other cases that would get messed up still though. I think a `code-block` would be safest.

**Comment by dbkinder** [comment]
We’re not using a code block today and haven’t noticed a problem.

-- david kinder

On Mar 28, 2018, at 12:48 PM, Ulf Magnusson <notifications@github.com<mailto:notifications@github.com>> wrote:


@ulfalizer commented on this pull request.

________________________________

In doc/scripts/genrest/genrest.py<https://github.com/zephyrproject-rtos/zephyr/pull/6843#discussion_r177868980>:

> +                      ".. title:: {0}\n\n"
+                      ".. option:: CONFIG_{0}\n\n"
+                      "{1}\n\n"
+                      "Type: ``{2}``\n\n"
+                      .format(sym.name,
+                              prompt_str,
+                              kconfiglib.TYPE_TO_STR[sym.type]))
+
+        # Symbols with multiple definitions can have multiple help texts
+        for node in sym.nodes:
+            if node.help is not None:
+                # Wrap the help text in a code block to preserve "ASCII art"
+                # and the like
+                sym_rst.write("Help\n"
+                              "====\n\n"
+                              ".. code-block:: none\n\n"


Talked a bit to @mbolivar<https://github.com/mbolivar> on IRC. It wouldn't take anything fancy for the formatting to get messed up either. Take a look at SYS_LOG_SPI_LEVEL in drivers/spi/Kconfig for example.

Looks like you can preserve newlines in RST by putting | at the beginning of each line. There are probably other cases that would get messed up still though. I think a code-block would be safest.

**Comment by ulfalizer** [comment]
@dbkinder 
Maybe it wasn't obviously broken if you didn't know what it was supposed to look like.

Here's how the old version of `genrest.py` renders the help text for `SYS_LOG_SPI_LEVEL` for example:
https://www.dropbox.com/s/kojxv0netilyaau/brokenhelp.png?dl=0 (all mangled up into a single paragraph)

Here's what it's supposed to look like (`drivers/spi/Kconfig`):

```
Sets log level for SPI drivers.
Levels are:
0 OFF, do not write
1 ERROR, only write SYS_LOG_ERR
2 WARNING, write SYS_LOG_WRN in addition to previous level
3 INFO, write SYS_LOG_INF in addition to previous levels
4 DEBUG, write SYS_LOG_DBG in addition to previous levels
```

Seems you can put a `|` at the beginning of each line in RST to force line breaks, which would fix that particular case, but that would still break stuff like alignment and (maybe) indentation.

**Review by mbolivar — APPROVED** [review]
I built this, looked at the results, and I think it's a solid improvement.

There are ugly scrollbars in the kconfig definitions for some options, but I think that's tolerable and can be fixed later. The benefit of the more appropriate data and otherwise improved formatting seems worth having.

**Comment by dbkinder** [comment]
Ah.  I see. I was trying to look at this with my phone while at the airport.  You can see the currently rendered output at http://docs.zephyrproject.org/reference/kconfig/CONFIG_SYS_LOG_SPI_LEVEL.html#cmdoption-arg-config-sys-log-spi-level

Yep.  The output is indeed just flattened out into one paragraph.  Your change to display the help as a fixed-width font code block seems overkill though.  My preference would be to keep the format as is today and then spend the time to edit the Kconfig files to improve how they're written so their output as HTML looks better.  (As stated above, fit it later.)

**Review by carlescufi — APPROVED** [review]
+1 to merging this

File: doc/scripts/genrest/genrest.py (hunk 1 above)
Comment (nashif) [reply doc/scripts/genrest/genrest.py]: I think the help text should be presented as text like it used to be, using code-block here looks really bad. The few cases were this get messed up we can live with or fix in Kconfig.

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by nashif** [comment]
image

with 
```
config SYS_LOG_SPI_LEVEL
        int
        prompt "SPI Driver Log level"
        depends on SYS_LOG && SPI
        default 0
        range 0 4
        help
          Sets log level for SPI drivers.  Levels are:

            0 OFF, do not write

            1 ERROR, only write SYS_LOG_ERR

            2 WARNING, write SYS_LOG_WRN in addition to previous level

            3 INFO, write SYS_LOG_INF in addition to previous levels

            4 DEBUG, write SYS_LOG_DBG in addition to previous levels

```

**Comment by ulfalizer** [comment]
I removed the `code-block`.

Piggybacking a formatting fix for `SYS_LOG_SPI_LEVEL` as well, which makes it consistent with `SYS_LOG_CRYPTO_LEVEL`. It now renders like this:

spihelp

**Comment by dbkinder** [comment]
Thanks!

-- david kinder

On Mar 29, 2018, at 8:52 AM, Ulf Magnusson <notifications@github.com<mailto:notifications@github.com>> wrote:


I removed the code-block.

Piggybacking a formatting fix for SYS_LOG_SPI_LEVEL as well, which makes it consistent with SYS_LOG_CRYPTO_LEVEL. It now renders like this:

[spihelp]<https://user-images.githubusercontent.com/154768/38099044-e3c04a98-3379-11e8-8bfd-34b375edfbf3.png>

**Review by nashif — APPROVED** [review]

**Pull request merged**